src/main/java/io/supertokens/totp/Totp.java

```diff
@@ -0,0 +1,301 @@
+
+        // FIXME: Should return base32 or base16
+        // Return base64 string of the secret key:
+        return Base64.getEncoder().encodeToString(keyGenerator.generateKey().getEncoded());
+    }
+
+    private static boolean checkCode(TOTPDevice device, String code) {
+        final TimeBasedOneTimePasswordGenerator totp = new TimeBasedOneTimePasswordGenerator(
+                Duration.ofSeconds(device.period));
+
+        byte[] keyBytes = Base64.getDecoder().decode(device.secretKey);
+        Key key = new SecretKeySpec(keyBytes, "HmacSHA1");
+
+        final int period = device.period;
+        final int skew = device.skew;
+
+        // Check if code is valid for any of the time periods in the skew:
+        for (int i = -skew; i <= skew; i++) {
+            try {
+                if (totp.generateOneTimePasswordString(key, Instant.now().plusSeconds(i * period)).equals(code)) {
+                    return true;
+                }
+            } catch (InvalidKeyException e) {
+                return false;
+            }
+        }
+
+        return false;
+    }
+
+    public static TOTPDevice registerDevice(Main main, String userId, String deviceName, int skew, int period)
+            throws StorageQueryException, DeviceAlreadyExistsException, NoSuchAlgorithmException {
+
+        TOTPSQLStorage totpStorage = StorageLayer.getTOTPStorage(main);
+
+        // Assert that period > 0 (not == 0 because it would lead to a divide by 0
+        // error)
+        // Assert that period <= 60. Otherwise, it is a security risk. Actually,
+        // anything > 30 is bad.
+        // and skew >= 0 and skew <= 2. Otherwise, it is a security risk.
+
+        // TODO: There should be a hard limit on number of devices per user
+        // 8 devices per user should be enough. Otherwise, it is a security risk.
+
+        String secret = generateSecret();
+        TOTPDevice device = new TOTPDevice(userId, deviceName, secret, period, skew, false);
+        totpStorage.createDevice(device);
+
+        return device;
+    }
+
+    private static void checkAndStoreCode(Main main, TOTPStorage totpStorage, String userId, TOTPDevice[] devices,
+            String code)
+            throws InvalidTotpException, StorageQueryException, TotpNotEnabledException, LimitReachedException {
+        // Note that here we are fetching all the codes (expired/non-expired).
+        // otherwise, because of differences in expiry time of different codes, we might
+        // end up with a situation where the will be released from the rate limiting too
+        // early because of some invalid codes in the checking window expired OR it can
+        // also lead to random rate limiting because if some valid codes blip out of the
+        // checking window and if it leads to N contagious invalid codes, then the user
+        // will be rate limited for no reason.
+        TOTPUsedCode[] usedCodes = totpStorage.getAllUsedCodes(userId);
```

Comment:
Ideally this should be a select for update cause theoritically, someone could send 100 requests in parallel with 100 different codes, and then this function would run with race conditions.

Reply:
How to do that here?